metrics/src/agent.rs

```diff
@@ -0,0 +1,266 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present Datadog, Inc.
+
+use crate::metrics::Metric;
+use crate::payload::construct_distribution_payload;
+
+use futures::lock::Mutex;
+use std::net::SocketAddr;
+use std::str;
+use std::sync::Arc;
+use std::time::UNIX_EPOCH;
+use std::time::{self, SystemTime};
+use tokio::net::UdpSocket;
+
+use hyper::header::CONTENT_TYPE;
+use hyper::http::HeaderValue;
+use hyper::{Body, Client, Request};
+use hyper_rustls::HttpsConnectorBuilder;
+
+const DOGSTATSD_PORT: u16 = 8125;
+const BUFFER_SIZE: usize = 8192;
+
+pub struct MetricsAgent {
+    config: Config,
+    buf: Arc<Mutex<Vec<Metric>>>,
+}
+
+impl MetricsAgent {
+    pub fn with_config(config: Config) -> Self {
+        Self {
+            config,
+            buf: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    pub async fn run(&self) {
+        // Create a UDP socket and bind to the dogstatsd port
+        let socket = UdpSocket::bind(
+            format!("0.0.0.0:{}", DOGSTATSD_PORT)
+                .parse::<SocketAddr>()
+                .unwrap(),
+        )
+        .await
+        .expect("Error binding to socket");
+
+        println!("Listening for dogstatsd packets on port {}", DOGSTATSD_PORT);
+
+        // We use a shared vec here instead of a channel because we need to buffer the data at
+        // some point anyways. If we had used a channel here, we'd still need to add it to
+        // some type of buffer, so we'd need to implement a vector buffer anyways.
+        // Instead, we can save a step and just add it to the buffer directly.
+        let buf_producer = self.buf.clone();
+        let buf_consumer = self.buf.clone();
+
+        let https = HttpsConnectorBuilder::new()
```

Comment:
not sure you need this, you can just use the builder

Reply:
The default client doesn't implement https, the docs recommend using an HttpsConnector for https requests. I think it's okay to leave it the way it is now.